api/pkg/client/mlflow/mlflow.go

```diff
@@ -0,0 +1,173 @@
+	if err != nil {
+		return err
+	}
+
+	if method == "POST" {
+		headers := map[string]string{
+			"Content-Type": "application/json",
+		}
+		for key, value := range headers {
+			req.Header.Set(key, value)
+		}
+	}
+
+	resp, err := mfc.Api.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if !(resp.StatusCode >= http.StatusOK && resp.StatusCode < http.StatusMultipleChoices) {
+		// Convert response body to Error Message struct
+		var errMessage DeleteExperimentErrorResponse
+		if err := json.NewDecoder(resp.Body).Decode(&errMessage); err != nil {
+			return err
+		}
+		return fmt.Errorf(errMessage.Message)
+	}
+
+	if response != nil {
+		if err := json.NewDecoder(resp.Body).Decode(response); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (mfc *mlflowClient) SearchRunForExperiment(idExperiment string) (SearchRunsResponse, error) {
+	// Search related runs for an experiment id
+	var responseObject SearchRunsResponse
+
+	searchRunURL := fmt.Sprintf("%s/api/2.0/mlflow/runs/search", mfc.Config.TrackingURL)
+
+	input := SearchRunRequest{ExperimentId: []string{idExperiment}}
+	jsonInput, err := json.Marshal(input)
+	if err != nil {
+		return responseObject, err
+	}
+
+	err = mfc.httpCall("POST", searchRunURL, jsonInput, &responseObject)
+	if err != nil {
+		return responseObject, err
+	}
+
+	return responseObject, nil
+}
+
+func (mfc *mlflowClient) SearchRunData(idRun string) (SearchRunResponse, error) {
+	// Creating Output Format for Run Detail
+	var runResponse SearchRunResponse
+	getRunURL := fmt.Sprintf("%s/api/2.0/mlflow/runs/get?run_id=%s", mfc.Config.TrackingURL, idRun)
+
+	err := mfc.httpCall("GET", getRunURL, nil, &runResponse)
+	if err != nil {
+		return runResponse, err
+	}
+	return runResponse, nil
+}
+
+func (mfc *mlflowClient) DeleteExperiment(idExperiment string) error {
+
+	relatedRunId, err := mfc.SearchRunForExperiment(idExperiment)
+	if err != nil {
+		return err
+	}
+	// Error Handling, keep runId that failed to delete
+	// TODO: What should i do with this?
+	var deletedRunId []string
+	var failDeletedRunId []string
+	for _, run := range relatedRunId.RunsData {
+		err = mfc.DeleteRun(run.Info.RunId, false)
+		if err != nil {
+			failDeletedRunId = append(failDeletedRunId, run.Info.RunId)
+			// return err
+		} else {
+			deletedRunId = append(deletedRunId, run.Info.RunId)
+		}
+	}
+
+	if len(relatedRunId.RunsData) > 0 {
+		// the [5:] is to remove the "gs://" on the artifact uri
+		// ex : gs://bucketName/path → bucketName/path
+		path := relatedRunId.RunsData[0].Info.ArtifactURI[5:]
+
+		// This section is getting gcs path for a folder, from a run id
+		// ex : bucketName/mlflow/idExperiment/runId/artefact  → bucketName/path/mlflow/idExperiment
+		splitPath := strings.SplitN(path, "/", 4)
+		folderPath := strings.Join(splitPath[0:3], "/")
+		// deleting folder
+		err = mfc.GcsService.DeleteArtifact(folderPath)
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
Hmmmm, i just realized that deleting a folder, or deleting a run after the soft deletion call the same number of deletion function from GCS, since it was iterating through all the folder, and delete each file separately.

How about if i implement this

DeleteExperiment
-- Get Every Related Run
-- Loop
-- -- Soft Delete Run
-- -- GCS Deletion Run

Then both DeleteRun, DeleteExperiment wont accept any bool parameter anymore, because all the gcs artefact will be deleted.

This implementation correspond with what happened if one of the run fail to get deleted. Referring to

> Good question, I believe we should just break the entire operation and return an error along the lines of deletion failed for run_id [id] for experiment [name]: [encapsulated error] and stop deleting the rest. I don't think there's an easy way to make this API call atomic, but it'll be nice to hear if others have an opinion on this :)

Reply:
> Yes, the zap logger API `Infow` works with KV pairs. No one is using the console logger in prod now, but this format may be preferable to `String()` if we end up enabling it at some point (eg: forward console logs to NewRelic).

Makes sense. I agree with this.

Reply:
> As for the implementation of this DeleteExperiment function, I was also thinking you can just loop and call DeleteRun directly right? 🤔

Yepp, using DeleteRun func to delete

Reply:
> > Yes, the zap logger API `Infow` works with KV pairs. No one is using the console logger in prod now, but this format may be preferable to `String()` if we end up enabling it at some point (eg: forward console logs to NewRelic).
> 
> Makes sense. I agree with this.

Do you mean the bool parameter in Delete Experiment and Delete Run right?

Reply:
How did this comment end up here? 🤣
https://github.com/gojek/mlp/pull/75#discussion_r1141588525

Reply:
How did this comment end up here? 🤣
https://github.com/gojek/mlp/pull/75#discussion_r1141588525